### ISSUE TITLE
Conventional documentation for env vars/globals?

### ISSUE THREAD

**Issue by jasonkarns** [docopt/docopt#437]
I absolutely love docopt! (Even when not leveraging docopt for actual parsing, I use the online tester for generating usage docs.)

Are there any known conventions for documenting global or environment variables that are used/respected by the CLI?

Along the same lines, are there any established conventions for how to document what is read on STDIN, or printed to STDERR/STDOUT?

**Comment by mentalisttraceur** [comment]
I don't know of a convention, but I've done a lot of thinking about this, and here's what I would use and propose as a convention:

## Environment Variables

Add an "Environment:" section to the help text, basically following a format which is a simplified version of the format for the "Options:" section. For example, a typical entry might be:

```
Environment:
    TEMPDIR  Directory for temporary files [default: /tmp]
```

Two spaces between environment variable name and description, both to match Docopt option descriptions, and to force things to be a little more readable in pathological cases where a bunch of environment variables are the same length:

```
Environment:
    FOO  Some description of foo
    BAR  Same as option --bar
    QUX  Whatever else
```

Note that by default it is okay to just do `FOO`, you don't have to do `FOO=<foo>` or `FOO=FOO` or `FOO=<>` like you would have to do with Docopt options if you had an option `--foo` that took an argument. This is because the nearly universal use for the environment variables is to pass in a value - it is fairly rare and a little confusing to a lot of users for the presence of an environment variable to work like a boolean flag the way that the presence of a command-line option often does. (Especially because you rarely actually want `FOO= whatever` to treat `FOO` as set, most people find it more intuitive for "set to empty string" to be equivalent to "not set", and sometimes people also expect `FOO=false` or `FOO=0` to work, so it's nicer to our users for us to explicitly consider and then decide exactly what is or isn't allowed.)

But the `FOO=...` syntax is still supported, precisely so that you can document non-arbitrary options and more complex usage patters. For example:

```
Environment:
    FOO=(true|false)  Control if foo is used
```
```
Environment:
    REMOTE=(false|0|no|off|)  Disable syncing with remote
    REMOTE=<FQDN or IP>       Address of the remote to sync
```

Also, this makes it easier to show when something can be specified in either an environment variable or an option:

```
Usage: foo [--bar=<bar>]

Options:
    --bar=<bar>  The bar with which to foo
    
Environment:
    BAR=<bar>  Default if option --bar is not given
```

---

## STDIN, STDOUT, STDERR, etc

Add an "Input/Output:" section to the help text. The format is very simple: the "name" of the input or output, two spaces, and the description. The "name" can be either:
 
1. `stdin`, `stdout`, `stderr`, `standard input`, `standard output`, or `standard error`,
2. one of the parameter names that appear in the "Usage", "Options", or "Environment" sections,
3. a relative or absolute file path (optionally with parameters from those other sections in it),
4. a URL (optionally with parameters from those other sections in it).

For example, here's how we might use this convention when documenting a very minimal `cat` implementation:

```
Usage: cat [<option>]... [--] [<file>]...

Options:
    --help     display this help and exit
    --version  output version information and exit

Input/Output:
    standard output  the catenated contents of each <file>
```

And here is a simple `tee`

```
Usage: tee [<option>]... [--] [<file>]...

Options:
    -a, --append             append to <file> instead of overwriting
    -i, --ignore-interrupts  ignore interrupt signals
    --help                   display this help and exit
    --version                output version information and exit

Input/Output:
    standard input   read until EOF and copied to each <file>
```

Note that in these two examples, having a separate line for `<file>` in the `Input/Output` section is redundant because we already said everything we need to say about how <file> arguments are used, but we might include it help tooling which understands this convention: imagine a shell which calls and parses `whatever --help` to determine what the autocompletion should be for arguments to `whatever`, or an IDE which calls and parses `whatever --help` to provide smarter checking of arguments when writing a shell script - then if we include a dedicated line for `<file>`, the shell can know to suggest file paths for that argument, and the IDE can warn if it seems that you're passing something else to it.

Programs like `bash` or `curl` could mention files like `~/.bashrc` or `~/.netrc` in this section.

Here's what I imagine a simple `nc` looking like in this scheme:

```
Usage:
    nc [-u] [-4|-6] <host> <port>
    nc -l [-4|6] <port>
    nc (-h | --help | -V | --version)

Options:
    -l, --listen   listen for connections instead of initiating them
    -u, --udp      use UDP instead of TCP
    -4, --ipv4     force using only IPv4 addresses
    -6, --ipv6     force using only IPv6 addresses
    -h, --help     print this help text
    -V, --version  print version information

Input/Output:
    standard input       data to send to the remote host
    standard output      data received from the remote host
    tcp://<host>:<port>  the remote when option -u is not given
    udp://<host>:<port>  the remote when option -u is given
```

I think it would also make sense to support "Input:" and "Output:" sections, to help reduce ambiguity or verbosity regarding which ones are just inputs, which ones are just outputs, and which ones are both. But I think in most cases, there's nothing to document besides maybe one or both of standard input standard output, and for those cases, a single "Input/Output" section is nicer for everyone involved.

---

Last I checked, `docopt` ignored all sections that it did not recognize, so these two proposals should be compatible with current versions of `docopt`.

**Comment by mentalisttraceur** [comment]
Also, you didn't ask, but I have a similar answer with proposed conventions for exit codes and error messages:

## Exit Codes

This section would be called "Exits:", and would document the meaning of the exit status that the program returns with. For example, a simple `diff` might use this section like this:

```
Exits:
    0  If the input files are the same
    1  If the input files differ
    2  If there was an error
```

If this section is omitted, the "standard" uncomplicated behavior is assumed: if the program does its job successfully, it returns the standard/conventional "success" exit value (`0` on most modern systems, an empty string on systems like Plan 9, etc), and if the program encountered an error, it returns any one of the standard/conventional "error" exit values (anything other than `0` on most modern systems, a non-empty string on systems like Plan 9, etc).

## Error Messages

This documents error messages.

The section is named "Errors:", and documents errors which your program might print on standard error. (On systems like Plan 9, where exit statuses are strings, this can also include the exit value, which is why on systems like Plan 9 the earlier "Exits:" section would be redundant pretty much all of the time).

The error messages are assumed to be self-descriptive, and thus not requiring further description. Here's an example that would be appropriate for one of my programs:

```
Errors:
    need option or directory argument
    bad option
    error writing output
    error changing directory
    error executing command
```

It is *implied* and *required* that all these error messages are prefixed with your program's zeroth argument (this is the name with which your program was invoked - if this is not available, either the canonical/official/intended name of your program, or the empty string, are to be used instead), a colon, and a single space, in that order.

It is *allowed* that your error message is followed by one or more error strings also separated by `": "`.

Basically, the idea is that if you're adding this section, you're committing to having error messages that look like this:

```
<program name>: <the documented error message>[: <other text>]...
```

because that way any user and code that calls your program can rely on specific error messages in a specific format that can be reliably parsed out with very simple code.

**Comment by mentalisttraceur** [comment]
One more convention proposal to the pile:

## Parameter Formats

Add a section named `Formats` to specify the valid format for one or more of the `<...>` placeholders in any of the sections.

Example for [the POSIX `umask` command](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/umask.html):

```
Usage:
    umask [-S] [<mask>]

Options:
    -S  show the umask symbolically instead of in octal

Format:
    <mask>         <octal> | <symbolic>[,<symbolic>]...
    <symbolic>     [u|g|o|a]...((-|+|=)[r|w|x]...)...
    <octal digit>  [0]...[[<octal digit>]<octal digit>]<octal digit>
    <octal digit>  0 | 1 | 2 | 3 | 4 | 5 | 6 | 7
```

Notice that we're only using the same format that Docopt already uses for usage strings, and that was enough to fully specify the full range of possible valid umask arguments.

This is valuable because often help text might say something like `<foo>` to indicate that you need to pass in a "foo", but

1. an unfamiliar user might have no idea what a "foo" looks like! is it one of a few valid strings? is it anything at all? etc,
2. a familiar user might need a quick reminder about some edge case,
3. all those amazing tools we could have thanks to parseable help text can't be amazing if they don't know what `<foo>` looks like,
and
4. if we use the same syntax `docopt` implementations can't help you resolve the boilerplate of parsing and validating that individual parameter string, even though they may have a lot of the necessary logic already just to parse the parameter list.

(The reason I propose naming this section `Format` is that it is more broadly familiar and less technically hairsplit and inconsistently used than words like `Syntax` and `Grammar`.)
